### PR TITLE
Security fixes: Limit number of keys learned, Limit key sizes learned, fixes in accordance to security review

### PR DESCRIPTION
**Pull request by davidhadas**
Fixes: #225 
Fixes: #243 

1. Added a mechanism to limit the number of learned keys and accumulate all remaining keys as unknown keys
2. Added a mechanism to limit the number of characters in keys inside keyvals - hash using xxHash when key is too long
3. ADA-KNATIVE-23-5 (remove profiling endpoints)
4. ADA-KNATIVE-23-6 (protect against `slowloris` - by adding `ReadHeaderTimeout`)
5. ADA-KNATIVE-23-12 (Added `Security.md`)
6. ADA-KNATIVE-23-13 (Reject requests with a large body /sync)
7. ADA-KNATIVE-23-14 (Reject requests with a large body /mutate)

```release-note
     Security-Guard added learning of unknown keys and consequently limited learned criteria accumulated size
     Security-Guard security fixs: fixed ADA-KNATIVE-23-5, ADA-KNATIVE-23-6, ADA-KNATIVE-23-12, ADA-KNATIVE-23-13, ADA-KNATIVE-23-14
```

### CODE REVIEW

File: pkg/guard-kubemgr/guard-kubemgr.go [hunk 1]
```diff
@@ -325,42 +333,51 @@
...
-	if g.Spec == nil {
-		g.Spec = new(spec.GuardianSpec)
+	learned, errMarshal := json.Marshal(guardianSpec.Learned)
+	if errMarshal != nil {
+		return fmt.Errorf("json.Marshal(guardianSpec.Learned) error for ns %s sid %s: %w", ns, sid, errMarshal)
 	}
-	if guardianSpec != nil {
-		if guardianSpec.Control != nil {
-			g.Spec.Control = guardianSpec.Control
+	str := fmt.Sprintf(`[{"op":"replace","path":"/spec/learned","value":%s},{"op":"replace","path":"/spec/samples","value":%d}]`, learned, guardianSpec.NumSamples)
```
[the remark this answers is not in the document]
Comment (davidhadas) [reply pkg/guard-kubemgr/guard-kubemgr.go:340]: https://github.com/github/codeql/issues/14159

**Comment by davidhadas** [comment]
Resetting EasyCLA by closing and reopening this PR

**Review by psschwei — COMMENTED** [review]
/lgtm

**Pull request merged**